CDCgov/prime-reportstream
# Automatically convert FHIR Dates and Times to HL7 format

This PR ...
When translating from FHIR back to HL7, automatically convert dates and times to the right format so that we don't have to add per field transforms in the schema. This converts FHIR Dates, Times, DateTimes, and Instants into their associated HL7 DTM or TM.

Relevant Specs:
- FHIR Datatypes: https://build.fhir.org/datatypes.html#primitive
- HL7 DTM: https://hl7-definition.caristix.com/v2/HL7v2.6/DataTypes/DTM
- HL7 TM: https://hl7-definition.caristix.com/v2/HL7v2.8/DataTypes/TM

Test Steps:
1.  Run `./prime fhirdata --input-file "src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir" -s "src/test/resources/fhirengine/translation/hl7/schema/schema-read-test-01/ORU_R01.yml"`
2. You should see a properly translated Timestamp in the MSH segment `MSH|^~\&|||||20280808112805.0000-0400||ORU^R01^ORU_R01|Bundle/df81fb05-e2aa-4538-8552-b8554bdea307||2.5.1`
3. You can also run the added unit tests to see all the different format types

## Changes
- Updates the evaluateString FhirPathUtils Function to use newly added converter methods
- Added unit tests for each of the methods introduced
- Added a timestamp to the FHIR test schema so that it can be tested using the primeCLI command

## Checklist

### Testing

### Process

## Linked Issues
- Fixes #issue

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
